network/transport/stcp_socket.c

```diff
@@ -0,0 +1,260 @@
+	/** Certification authority certificate */
+	mbedtls_x509_crt		ca_cert[1];
+	/** Client certificate */
+	mbedtls_x509_crt		client_cert[1];
+	/** Client private key */
+	mbedtls_pk_context		client_pkey[1];
+	/** SSL configuration strucure */
+	mbedtls_ssl_config		conf[1];
+	/** Mbedtls tls context */
+	mbedtls_ssl_context		ssl[1];
+};
+
+static int32_t stcp_socket_init(struct stcp_socket_desc **desc,
+				struct network_interface *net,
+				struct stcp_socket_init_param *param);
+static int32_t stcp_socket_remove(struct stcp_socket_desc *desc);
+static int32_t stcp_socket_connect(struct stcp_socket_desc *desc,
+				   struct socket_address *addr);
+static int32_t stcp_socket_disconnect(struct stcp_socket_desc *desc);
+static int32_t stcp_socket_send(struct stcp_socket_desc *desc, const void *data,
+				uint32_t len);
+static int32_t stcp_socket_recv(struct stcp_socket_desc *desc, void *data,
+				uint32_t len);
+
+/* Secure socket interface */
+const struct socket_interface stcp_socket_interface = {
+	.protocol = PROTOCOL_TCP,
+	/* Will be initialized in socket.c */
+	.sock = NULL,
+	.init = (int32_t (*)())stcp_socket_init,
+	.remove = (int32_t (*)())stcp_socket_remove,
+	.connect = (int32_t (*)())stcp_socket_connect,
+	.disconnect = (int32_t (*)())stcp_socket_disconnect,
+	.send = (int32_t (*)())stcp_socket_send,
+	.recv = (int32_t (*)())stcp_socket_recv,
+	/* Only for PROTOCOL UDP */
+	.sendto = NULL,
+	.recvfrom = NULL
+};
+
+/* Wrapper over socket_recv */
+static int tls_net_recv(struct socket_desc *sock, unsigned char *buf,
+			size_t len)
+{
+	int32_t ret;
+
+	ret = socket_recv(sock, buf, len);
+	if (ret == -EAGAIN)
+		return MBEDTLS_ERR_SSL_WANT_READ;
+
+	return ret;
+}
+
+/** @brief See \ref socket_init */
+static int32_t stcp_socket_init(struct stcp_socket_desc **desc,
+				struct network_interface *net,
+				struct stcp_socket_init_param *param)
+{
+	struct socket_init_param	sock_param;
+	struct stcp_socket_desc		*ldesc;
+	int32_t				ret;
+
+	if (!desc || !param || !param->ca_cert)
+		return FAILURE;
+
+	ldesc = (typeof(ldesc))calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return FAILURE;
+
+	/* Initialize structures */
+	mbedtls_x509_crt_init(ldesc->ca_cert);
+	mbedtls_x509_crt_init(ldesc->client_cert);
+	mbedtls_pk_init(ldesc->client_pkey);
+	mbedtls_ssl_config_init(ldesc->conf);
+
+	sock_param.type = TCP_SOCKET;
+	sock_param.net = net;
+	sock_param.init_param.tcp = NULL;
+	ret = socket_init(&ldesc->sock, &sock_param);
+	if (IS_ERR_VALUE(ret)) {
+		ldesc->sock = NULL;
+		goto exit;
+	}
+
+	ret = trng_init(&ldesc->trng, param->trng_init_param);
+	if (IS_ERR_VALUE(ret)) {
+		ldesc->trng = NULL;
+		goto exit;
+	}
+
+	/* Set default configuration: TLS client socket */
+	ret = mbedtls_ssl_config_defaults(ldesc->conf,
+					  MBEDTLS_SSL_IS_CLIENT,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+
+	/* Certificate validation is set as mandatory */
+	//TODO
```

Comment:
remove this TODO, also the commented code.

Reply:
done